include/dd/Node.hpp

```diff
@@ -33,39 +32,79 @@ namespace dd {
+                flags = (flags | 16);
+            else
+                flags = (flags & (~16));
+        }
+        inline void setSymmetric(bool symmetric) {
+            if (symmetric)
+                flags = (flags | 32);
+            else
+                flags = (flags & (~32));
+        }
     };
     using mEdge       = Edge<mNode>;
     using mCachedEdge = CachedEdge<mNode>;
 
-    inline mNode mNode::terminalNode{{{{nullptr, Complex::zero}, {nullptr, Complex::zero}, {nullptr, Complex::zero}, {nullptr, Complex::zero}}}, nullptr, 0U, -1, 0, true, true};
+    inline mNode mNode::terminalNode{{{{nullptr, Complex::zero}, {nullptr, Complex::zero}, {nullptr, Complex::zero}, {nullptr, Complex::zero}}}, nullptr, 0U, -1, 32 + 16};
 
     struct dNode {
         std::array<Edge<dNode>, NEDGE> e{};    // edges out of this node
         dNode*                         next{}; // used to link nodes in unique table
         RefCount                       ref{};  // reference count
         Qubit                          v{};    // variable index (nonterminal) value (-1 for terminal)
-        std::int_fast8_t               flags = 0;
+        std::uint_least8_t             flags = 0;
+        // 32 = marks a node with is symmetric.
+        // 16 = marks a node resembling identity
         // 8 = marks a reduced dm node,
         // 4 = marks a dm (tmp flag),
         // 2 = mark first path edge (tmp flag),
         // 1 = mark path is conjugated (tmp flag))
 
         static dNode            terminalNode;
         constexpr static dNode* terminal{&terminalNode};
-
-        static constexpr bool isTerminal(const dNode* p) { return p == terminal; }
-        //static constexpr bool isDensity() { return true; }
+        static constexpr bool   isTerminal(const dNode* p) { return p == terminal; }
     };
     using dEdge       = Edge<dNode>;
     using dCachedEdge = CachedEdge<dNode>;
     inline dNode dNode::terminalNode{{{{nullptr, Complex::zero}, {nullptr, Complex::zero}, {nullptr, Complex::zero}, {nullptr, Complex::zero}}}, nullptr, 0, -1, 0};
 
+    template<>
+    constexpr bool dEdge::operator==(const dEdge& other) const {
+        assert(p != nullptr && other.p != nullptr);
+        return p == other.p && tempDensityMatrixFlagsEqual(p->flags, other.p->flags) && w.approximatelyEquals(other.w);
+    }
+
 } // namespace dd
+
+namespace std {
+    template<>
+    struct hash<dd::dEdge> {
+        std::size_t operator()(dd::dEdge const& e) const noexcept {
+            auto h1 = dd::murmur64(reinterpret_cast<std::size_t>(e.p));
+            auto h2 = std::hash<dd::Complex>{}(e.w);
+            assert((dd::dEdge::isDensityMatrix((long)e.p)) == false);
+            auto h3 = std::hash<std::uint_least8_t>{}(dd::dEdge::returnTempDensityMatrixFlags(e.p->flags));
+            //           auto h3  = std::hash<short int>{}(e.p->flags);
+            auto tmp = dd::combineHash(h1, h2);
+            return dd::combineHash(tmp, h3);
+        }
+    };
```

Comment:
also here. just overload the hash for the `dNode` class and the hash for the `dEdge` class automatically works.

Reply:
Done. 
Note that the `dNode` Hash is necessary so that the hashes between a reduced matrix node (i.e., a density matrix node) and a matrix node are different. If this is not the case it comes to a mixup between the two node types.